src/game/crossteaser/mod.rs

```diff
@@ -53,31 +53,586 @@ const ABOUT: &str = "PLACEHOLDER";
+/// clever way.
+const TRANSFORM_MAP: [u64; 24] = [
+    0b0,             // 0
+    0b000_01,        // 1
+    0b000_11,        // 2
+    0b000_10,        // 3
+    0b010_01,        // 4
+    0b010_01_001_01, // 5
+    0b000_11_010_01, // 6
+    0b000_10_011_11, // 7
+    0b010_01_000_01, // 8
+    0b000_10_001_11, // 9
+    0b000_10_001_11, // 10
+    0b000_11_001_11, // 11
+    0b100_11,        // 12
+    0b100_11_011_01, // 13
+    0b100_11_010_11, // 14
+    0b000_10_001_01, // 15
+    0b000_11_010_11, // 16
+    0b010_11,        // 17
+    0b000_10_011_01, // 18
+    0b000_01_011_01, // 19
+    0b100_10_101_01, // 20
+    0b010_10,        // 21
+    0b001_10,        // 22
+    0b000_01_011_10, // 23
+];
+
+// Constant bitmask values that will be commonly used for hashing/unhashing
+// game states.
+const FACE_BITS: u64 = 3;
+const FACE_BITMASK: u64 = 0b111;
+const PIECE_SIZE: u64 = 24;
+
+/// Converts an Orientation struct into its corresponding orientation hash,
+/// which will be a number from 0-23
+/// Uses patterns in the binary representations of an orientation to generate
+/// a unique & minimal hash
+/// ~1.85x faster than indexing into an array like unhash_orientation()
+/// Maybe I can figure out an efficient reverse function at some point :)
+fn hash_orientation(o: &Orientation) -> u64 {
+    return (o.front << 2) | ((o.top & 1) << 1) | (o.right & 1);
+}
+
+/// Converts an orientation hash into an Orientation struct
+/// Makes use of ORIENTATION_UNMAP for the conversion
+fn unhash_orientation(h: u64) -> Orientation {
+    let packed: u64 = ORIENTATION_MAP[h as usize];
+    return Orientation {
+        front: packed >> (FACE_BITS * 2),
+        top: packed >> FACE_BITS & FACE_BITMASK,
+        right: packed & FACE_BITMASK,
+    };
+}
+
+impl UnhashedState {
+    fn deep_copy(&self) -> UnhashedState {
+        let mut new_pieces: Vec<Orientation> = Vec::new();
+        for o in &self.pieces {
+            new_pieces.push(Orientation {
+                front: o.front,
+                top: o.top,
+                right: o.right,
+            });
+        }
+        return UnhashedState {
+            pieces: new_pieces,
+            free: self.free,
+        };
+    }
+}
+
+impl Session {
+    /// Returns the total number of game pieces in the current game variant
+    /// based on Session
+    fn get_pieces(&self) -> u64 {
+        return self.length * self.width - self.free;
+    }
+
+    /// Simple hash function that converts a vector of piece
+    /// orientations and an empty space represented by an integer into a 64 bit
+    /// integer (State) which uniquely represents that state.
+    /// Uses minimal space for all theoretical states, does not optimize
+    /// for obtainable states. There will be a lot of unused hashes here
+    /// because fewer than half of the theoretical states are obtainable in
+    /// 3x3 crossteaser.
+    fn hash(&self, s: &UnhashedState) -> State {
```

Comment:
Since I recently changed these to use a base 24 representation in order to reduce maximum hash size, I'm not sure bitvec would work great here. I could move back to the original 5 bits per piece hash, it would probably be a lot faster than repeatedly multiplying/dividing by 24, but the tradeoff is much larger hash values.

Reply:
That makes sense, thanks!